README.md
[![](https://instagram.bot.ptkdev.io/img/instagrambot_logo.png)](https://instagram.bot.ptkdev.io)

# InstagramBot.js

[![](https://img.shields.io/badge/license-GLPv3-brightgreen.svg)](#) [![](https://img.shields.io/badge/powered%20by-puppeteer-46aef7.svg)](https://github.com/GoogleChrome/puppeteer) [![](https://img.shields.io/badge/version-v0.9.2-lightgrey.svg)](https://github.com/social-manager-tools/instagram-bot.js/releases) [![](https://img.shields.io/badge/chat%20on-slack-orange.svg)](http://slack.ptkdev.io) [![](https://img.shields.io/badge/chat%20on-discord-7289da.svg)](http://discord.ptkdev.io) [![](https://img.shields.io/badge/blog-medium-2AE176.svg)](http://blog.ptkdev.io) [![](https://img.shields.io/badge/twitter-ptkdevio-2AA3EF.svg)](https://twitter.com/ptkdevio)

[![](https://img.shields.io/badge/donate-patreon-F87668.svg)](http://patreon.ptkdev.io) [![](https://img.shields.io/badge/donate-paypal-46AFE0.svg)](http://paypal.ptkdev.io) [![](https://img.shields.io/badge/buy%20me-coffee-4B788C.svg)](http://coffee.ptkdev.io) [![](https://img.shields.io/badge/help-support@ptkdev.io-fbbc05.svg)](mailto:support@ptkdev.io)

[![](https://ptkdev.it/img/bot/ptkdev-instagram-bot.gif)](https://instagram.bot.ptkdev.io)

## What it does
This bot helps you increase the engagement level of your Instagram profile through different social algorithms.
Increase the likes on your photos and followers!

## Features
* [✓] Easy to use
* [✓] Login
* [✓] 2FA (bad location)
* [✓] 2FA (sms pin enabled)
* [✓] Multi-Session
* [✓] Multi-Platform (Windows 10, Mac OSX, Linux, [Raspberry PI 3](https://github.com/social-manager-tools/instagram-bot.js/blob/master/INSTALL.md))
* [✓] Error management feature (bad pin, bad password)
* [✓] Screenshot and Verbose logger
* [✓] Like Mode Classic: bot selects a random hashtag from a config list and likes 1 random photo, and can repeat this all time.
* [✓] Like Mode Realistic: bot selects a random hashtag from a config list and likes fast 10-12 photos, it sleeps 15-20min and repeats this all time. Sleeps at night.
* [✓] Like Mode Competitor Users: it selects an account, selects random followers, likes 10-12 photo and sleeps 15-20min. Sleeps at night.
* [✓] Like Mode Superlike: it selects random hashtag from a config list and likes 3 random photos of the same user.
* [✓] Comment Mode Classic: it selects random comments and random hashtags and writes comments under photos.
* [✓] Follow/Defollow Classic: follow 30 users, defollow first and rotate (in loop). This method is not detect from socialblade. ~1h | 300 follow-defollow/day.

## Fast setup (CLI Version)
1. Download [stable bot version](https://github.com/social-manager-tools/instagram-bot.js/releases) and extract it.
2. Download [Node.js](https://nodejs.org/it/) >= 7.6 and install it.
3. Run `npm install` in `instagram-bot.js` folder.
4. Rename `configs/config.js.tpl` to `configs/config.js`, fill it properly.
5. Start the bot via `node bot.js`
6. If it works add a star :star: at this project :heart:
7. If you want to help me: **[donate on paypal](http://paypal.ptkdev.io)** or become a **[backer on patreon](http://patreon.ptkdev.io)**.

For advanced configuration see [INSTALL.md](https://github.com/social-manager-tools/instagram-bot.js/blob/master/INSTALL.md).

#### 2FA: SMS Pin
If you received an SMS or an email pin edit `loginpin.txt` and insert it on the first line. Wait 50-60 seconds...

#### Tips: hide browser
Edit `configs/config.js` and switch `chrome_headless` option to `true`.

#### Check if it works:
See images in `./logs/screenshot` or disable `chrome_headless` flag.

#### Parameters
Run bot with different configs (multi account), use: `node bot.js --config="./configs/config.js"`

| Name     | Type    | Default                | Description                                  |
| ---      | ---     | ---                    | ---                                          |
| config   | String  | ./configs/config.js    | File path of the option file to run          |

## Bugs
1. `[ERROR] login: The username you entered doesn't belong to an account. Please check your username and try again. (restart bot and retry...)`
* Why did it happen? Instagram desktop is in overcapacity. Happens at 12:00-14:00 and 19:00-21:00 every day.
* Solution: Try to Login at a different time or Logout from your instagram app, and Login again. Reboot the bot and retry... Try and retry, and retry, and retry... Or stop the bot and wait 2-3h.

2. `Error: Protocol error (Page.captureScreenshot): Target closed.`
* Why did it happen? macOS doesn't properly support screenshots from puppeteer
* Solution: set `screenshot` as `false` in `configs/config.js`.

3. `This code is no longer valid. Please request a new one. (400) (/accounts/login/ajax/two_factor/)`
* Why did it happen? It's an Instagram bug at Login.
* Solution: disable momentarily the "2FA" or try an old version of chrome (edit `config.js` set `executable_path`)

## Desktop setup (GUI Version)
1. Download [Social Manager Tools GUI](https://socialmanagertools.ptkdev.io/).
2. Run application.

## Docker setup
If you prefer to run this using Docker, an official container is available from the [Docker Hub](https://hub.docker.com/r/socialmanagertools/instagram-bot.js).

In order to run it, copy the `config.js.tpl` file, configure it as you prefer, then use it through volume mapping,
like in this example:

```sh
$ docker run \
    --restart=always \
    --name=instagram-bot \
    -d \
    -v /path/to/config.js:/app/configs/config.js \
    socialmanagertools/instagram-bot.js &>/dev/null
```

## Roadmap
See full roadmap (open task, todo and bugs) in [project page](https://github.com/social-manager-tools/instagram-bot.js/projects?query=is%3Aopen+sort%3Aname-asc).
* [v0.9.X](https://github.com/social-manager-tools/instagram-bot.js/projects/3)
* [v1.0.X](https://github.com/social-manager-tools/instagram-bot.js/projects/3)

## Sorry for snake_case
I love :snake: snake_case syntax sorry for this :sob: don't hate me.

[![](https://socialmanagertools.ptkdev.io/img/socialmanagertools_logo.png)](https://github.com/social-manager-tools)

# Social Manager Tools

[Social Manager Tools GUI](https://github.com/social-manager-tools/social-manager-tools)  
[InstagramBot.js](https://github.com/social-manager-tools/instagram-bot.js) ([LIB](https://github.com/social-manager-tools/instagram-bot-lib))  
[TwitterBot.js](https://github.com/social-manager-tools/twitter-bot.js) ([LIB](https://github.com/social-manager-tools/twitter-bot-lib))  
[FacebookBot.js](https://github.com/social-manager-tools/facebook-bot.js) ([LIB](https://github.com/social-manager-tools/facebook-bot-lib))  
[WordpressTelegramBot.js](https://github.com/social-manager-tools/wordpress-telegram-bot.js) ([LIB](https://github.com/social-manager-tools/wordpress-telegram-bot-lib))  
[MediumTelegramBot.js](https://github.com/social-manager-tools/medium-telegram-bot.js) ([LIB](https://github.com/social-manager-tools/medium-telegram-bot-lib))  

# License

GNU GENERAL PUBLIC LICENSE

Copyright (c) 2018 Patryk Rzucidło (PTKDev)


CHANGELOG.md
# v0.9.2 (15 September, 2018)
* Downgrade: puppeteer v1.4.0 (#27)


# v0.9.1 (15 September, 2018)
* Fix: Node is either not visible or not an HTMLElement (#82 #86 #84)
* Fix: Random crash of Follow/Defollow Mode Classic
* Refactor: 2FA flow
* Update: New user-agent
* Upgrade: puppeteer v1.8.0


# v0.9.0b (09 September, 2018)
* Fix: bad creation of folders `databases` and `logs` (#79)


# v0.9.0 (08 September, 2018)
* Feature: better multi-account with `--config="./configs/config.js"` argv
* Feature: Follow/Defollow Mode (#77)
* Feature: Support sqllite connector (#70)
* Fix: Like Realistic random freeze
* Fix: Comment Mode don't work
* Fix: Like Competitor Mode (#76)
* API: stop() | stop the bot from your application
* Update: bot_fastlike_min and max from config.js now all mode use bot_likeday_min and max
* Update: bot_sleep_night disabled at default
* Update: `config.js.tpl` moved to `configs` folder
* Upgrade: puppeteer v1.7.0


# v0.8.2 (26 July, 2018)
* Fix: default path of logs in `config.js.tpl`


# v0.8.1 (25 July, 2018)
* Fix: Unliking already liked pictures (all mode) [Bug: #59]
* Fix: Commenting already commented pictures
* Update: `likemode_realistic` now is default mode in `config.js.tpl`
* Update: `900` now is default value of max like/day in `config.js.tpl`


# v0.8.0 (17 July, 2018)
* Fix: likemode_superlike don't like
* Feature: core moved to [instagrambotlib](https://www.npmjs.com/package/instagrambotlib)


# v0.7.5 (03 July, 2018)
* Fix: instagram selectors css changed [Bug: #55]


# v0.7.4 (28 June, 2018)
* Update: RaspBerry PI `INSTALL.md`
* Fix: random stop of `likemode_superlike`
* Feature: check updates of bot
* Feature: donate message
* Feature: default `config.js` value if don't exist [Bug: #51]


# v0.7.3 (11 June, 2018)
##### NOTE: Remove `node_modules` and run `npm install`.
* Fix: bot don't work with `puppeteer v1.5.0` package.json moved to `1.4.0`


# v0.7.2 (07 June, 2018)
##### NOTE: Update your `config.js` with new parameters from `config.js.tpl`.
* Fix: commentmode_classic `.state error`
* Feature: enable/disable sleep night
* Feature: likemode_superlike
* Feature: likemode_superlike, number of like/user from config.js
* Fix: sms pin bad location ust be inserted within 50-60 seconds


# v0.7.1 (02 June, 2018)
##### NOTE: Update your `config.js` with new parameters from `config.js.tpl`.
* New instagram limit: now the sms pin must be inserted within 50-60 seconds
* Fix and Refactor: commentmode_classic
* Fix and Refactor: competitor_users
* Fix and Refactor: likemode_realistic
* Fix and Refactor: likemode_classic
* Fix and Refactor: login
* Fix and Refactor: 2FA
* Refactor: logger
* Refactor: utils
* Refactor: state


# v0.7.0 HOTFIX (31 May, 2018)
* HotFix: instagram selectors changed, you need update to v0.7. InstagramBot.js <= v0.6.1 don't work.
* Feature: [Raspberry PI 3 Support](https://github.com/social-manager-tools/instagram-bot.js/blob/master/INSTALL.md)
* Feature: commentmode_classic - select random comment and random hashtag and write comment under photo.
* Refactor: removed useless comments, refactor switch, logger and constants.
* Refactor: snake_code

NOTE: This release don't respect full [roadmap v0.7](https://github.com/social-manager-tools/instagram-bot.js/projects/1), is a hot fix.


# v0.6.1 (February 14, 2018)
* `config.js.tpl` misprint, please rename one of `bot_fastlike_min` to `bot_fastlike_max`
* likemode_realist: fixbug, now is stable. If you need more like without soft ban set `bot_fastlike_min` 12 and `bot_fastlike_max` to 16 in `config.js`


# v0.6.0 (February 10, 2018)
* Feature: likemode_realistic. Update `config.js` from `config.js.tpl`
* Feature: Docker available (Thanks [@julianxhokaxhiu](https://github.com/julianxhokaxhiu/))
* Fix: error promise crash after 34-35h no-stop. 
* Fix: package name not support "dot" in name in last npm version
* Refactor: change my email address
* News: Available in Social Manager Tools the [Telegram Bot](https://github.com/social-manager-tools/wordpress-telegram-bot.js) for any blog Wordpress.


# v0.5.3 (January 31, 2018)
* Update documentation (comments in code)
* Warning: update max like value in `config.js.tpl` instagram in 2018 descreased like/day now max is 600 not 1000
* Feature: warning log message if hashtag has no photos
* Feature: ISSUE template for bug reports
* Bugs: bot does not work on Raspberry Pi because rasbian have chromium v60 (mininium version required is v63 and node v7.6)


# v0.5.2 (January 28, 2018)
* Fix [[#3](https://github.com/ptkdev/instagram-bot.js/issues/3)]: Like doesn't work if the language is not in English


# v0.5.1 (January 27, 2018)
* Fix: package.json syntax error


# v0.5.0 (January 26, 2018)
* New logo! Thanks to @maXfer75 and @TaniK72 on twitter.
* Bye bye webdriverio and selenium: moved to official google library "puppeteer". Better ram usage, better stability and no java required. Dropped firefox, ie and opera support (only chrome). Added multiplatform support (mac, windows, linux). INFO: The webdriverio old version of bot (0.4.2) is available on branch "webdriverio".


# v0.4.2 (January 25, 2018)

* Refactor: moved to selenium npm package


# v0.4.1 (January 23, 2018)

* Feature: like/day in config (max and min)
* Feature: method pin
* Refactor: removed self (useless)
* Fix: missing apos in config


# v0.4.0 (January 21, 2018)

* Refactor: async wait v2
* Refactor: bot.js without function flow
* Fix: clear array if # not have photos
* Fix: java heap size error (now is max 2gb)
* Fix: errors 2FA location 


# v0.3.1 (January 18, 2018)

* Fix: 110% cpu usage
* Fix: flow without pin


# v0.3.0 (January 18, 2018)

* Refactor: moved to async wait
* Refactor: add pm2 and forever to npm run
* Update: chromedriver binary
* Fix: get url picture selector
* Fix: <3 selector
* Fix: screen server resolution
* Fix: 2FA-security location
* Fix: 2FA-security enabled


# v0.2.1 (January 15, 2018)

* Ops: bad run script in package.json


# v0.2.0 (January 15, 2018)

bot.js
/**
 * InstagramBot.js
 * =====================
 * Instagram Bot made with love and nodejs
 *
 * @author:     Patryk Rzucidlo [@ptkdev] <support@ptkdev.io> (https://ptkdev.it)
 * @file:       bot.js
 * @version:    0.9.0
 *
 * @link: https://github.com/GoogleChrome/puppeteer
 * @link: https://www.npmjs.com/package/instagrambotlib
 *
 * @license:    Code and contributions have 'GNU General Public License v3'
 *              This program is free software: you can redistribute it and/or modify
 *              it under the terms of the GNU General Public License as published by
 *              the Free Software Foundation, either version 3 of the License, or
 *              (at your option) any later version.
 *              This program is distributed in the hope that it will be useful,
 *              but WITHOUT ANY WARRANTY; without even the implied warranty of
 *              MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *              GNU General Public License for more details.
 *              You should have received a copy of the GNU General Public License
 *              along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @link        Homepage:     https://instagram.bot.ptkdev.io
 *              GitHub Repo:  https://github.com/social-manager-tools/instagram-bot.js
 */
const argv = require("yargs").argv;
const dotenv = require("dotenv").load();
const fs = require("fs");
const Bot = require("instagrambotlib");

const config_file = (argv.config ? argv.config : "./configs/config.js");

if (fs.existsSync(config_file)) {
    let bot = new Bot(require(config_file));
    bot.start();
} else {
    throw new Error("Config file not found");
}


configs/config.js
module.exports = {
    // InstagramBot.js Configs
    "debug": true,
    "login": true,
    "ui": false, // only for social-manager-tools

    // Instagram Account
    "instagram_username": process.env.IG_LOGIN , // without @
    "instagram_password": process.env.IG_PW ,
    "instagram_hashtag": ["gotitforcheap", "ChicagoArtBookFair", "johnmaus", "posterdesign", "highsnobiety", "netart", "vaporwave", "babycore", "onbooooooom", "lebenliebegrafik", "arielpink", "riso", "risograph", "velvetropes", "postinternet", "itsnicethat", "herblubalin", "surrealism", "mirrorart", "2ddesign", "artdirection", "paulrand", "wimcrouwel", "yalegraphicdesign", "memphisgroup"], //without #, raccomanded more than 20
    "instagram_pin": "sms", //method to receive pin (email or sms)

    // BOT Configs
    // [WORK] likemode_classic          - select random hashtag from config list and like 1 random photo (of last 20) | 400-600 like/day.
    // [WORK] likemode_realistic        - select random hashtag from config list, like fast 10-12 photo and sleep 15-20min. Sleep at night
    // [WORK] likemode_superlike        - select random hashtag from config list and like 3 random photo of same user | 400-600 like/day.
    // [WORK] likemode_competitor_users - select account, select random followers, like 10-12 photo and sleep 15-20min.
    // [WORK] comment_mode              - leave a comment under the photo via hashtag.
    // [WORK] fdfmode_classic           - follow 30 users, and defollow the first followed at 31 follow (in loop). This method is not detected from socialblade. ~1h | 300/300 follow-defollow/day.
    "bot_mode": "fdfmode_classic",
    "bot_likeday_min": 800,      // work only in likemode_classic
    "bot_likeday_max": 900,      // work only in likemode_classic. In 2018 limit instagram is 1000/day
    "bot_sleep_night": false,    // if you need stop bot in night 
    "bot_start_sleep": "7:00",   // sleep from 00:00 to 7:00 am, work in likemode_realistic, comment_mode, likemode_superlike and likemode_competitor_users
    "bot_superlike_n": 3,        // superlike number of likes for user
    "bot_followday": 400,        // number of follow-defollow user/day, work in fdfmode_classic. In 2018 limit instagram is 500/day
    "bot_followrotate": 39,      // follow 30 users, and defollow the first followed at 31 follow (in loop), work in fdfmode_classic | If is 0 defollow all followed previously by bot
    "bot_userwhitelist": ["user1", "user2", "user3"],   // usernames without @ | don't defollow this users on follow-defollow mode, work in fdfmode_classic
    "comment_mode": {
        "comments": {
            "type": "array",
            "source": [": )", ":)", ";)", "⊂((・⊥・))⊃", "✌︎('ω')✌︎", "( ͡° ͜ʖ ͡°)", "✌︎('ω'✌︎ )", "(　ﾟдﾟ)"],
        }
    },
    "likemode_competitor_users": {
        "account": "user_nickname" // without @
    },

    // Puppeteer Configs
    "chrome_headless": true,
    "chrome_options": ["--disable-gpu", "--no-sandbox", "--window-size=1920x1080"],
    "executable_path": "", // example for Mac OS: /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome

    // DATABASES
    "fdfdatabase_path": "./databases/fdf.db", // databases with username of followed users with follodefollow mode
    "logdb_path": "./databases/logs.db",      // logs in sqllite database for better report

    // LOG
    "pin_path": "./loginpin.txt",
    "log_path": "./logs/debug.log",
    "logerr_path":"./logs/errors.log",
    "screenshot_path": "./logs/screenshot/",
    "log": {
        "drivers": ["console"], // slack or console
        "screenshot": false,  // disable or enable screenshot in logs folder
        "channels": {
            "console": "",
            "slack": {
                "webhook": ""
            }
        }
    }
};
